main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,242 @@
+import ca.uwaterloo.flix.util.ParOps
+
+/**
+  * The occurrence analyzer collects information on variable and function usage and calculates the weight of the expressions
+  * Marks a variable or function as Dead if it is not used, Once if it is used exactly once and Many otherwise
+  */
+object OccurrenceAnalyzer {
+
+  private object OccurInfo {
+
+    /**
+      * Occurrence information for an empty sequence of expressions
+      */
+    val Empty: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 0)
+
+    /**
+      * The initial occurrence information for an expression of size 1, i.e. an expression without subexpressions.
+      */
+    val One: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 1)
+  }
+
+  /**
+    * The occurrence of `defs` and `vars` inside the body of a `def`
+    * `size` represents the number of expressions in the body of a `def`
+    */
+  case class OccurInfo(defs: Map[DefnSym, Occur], vars: Map[VarSym, Occur], localDefs: Int, size: Int) {
+    def :+(kv: (DefnSym, Occur)): OccurInfo = {
+      this.copy(defs = this.defs + kv)
+    }
+
+    def +(kv: (VarSym, Occur)): OccurInfo = {
+      this.copy(vars = this.vars + kv)
+    }
+
+    def -(varSym: VarSym): OccurInfo = {
+      this.copy(vars = this.vars - varSym)
+    }
+
+    def --(varSyms: Iterable[VarSym]): OccurInfo = {
+      this.copy(vars = this.vars -- varSyms)
+    }
+
+    def get(defnSym: DefnSym): Occur = {
+      this.defs.getOrElse(defnSym, Dead)
+    }
+
+    def get(varSym: VarSym): Occur = {
+      this.vars.getOrElse(varSym, Dead)
+    }
+
+    def incrementLocalDefCount: OccurInfo = {
+      this.copy(localDefs = localDefs + 1)
+    }
+  }
+
+  private def increment(occurInfo: OccurInfo): OccurInfo = {
+    occurInfo.copy(size = occurInfo.size + 1)
+  }
+
+  /**
+    * A set of functions that contain masked casts.
+    *
+    * Must be manually maintained since Lowering erases the masked cast.
+    */
+  private val DangerousFunctions: Set[String] = Set("bug!", "Fixpoint.Debugging.notifyPreSolve", "Fixpoint.Debugging.notifyPostSolve", "Fixpoint.Debugging.notifyPreInterpret", "Assert.eq")
+
+  private def toReadableFunction(sym: Symbol.DefnSym): String = {
+    sym.toString.takeWhile(c => c.toString != Flix.Delimiter)
+  }
+
+  /**
+    * Performs occurrence analysis on the given AST `root`.
+    */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    val defs = visitDefs(root.defs) // visitDefs calls parMap internally and has additional handling
+    OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+  }
+
+  /**
+    * Performs occurrence analysis on every entry in `defs0` in parallel.
+    * Decorates each Def with occurrence information, i.e., how it appears in the program or if it is unused.
+    */
+  private def visitDefs(defs0: Map[DefnSym, OccurrenceAst.Def])(implicit flix: Flix): Map[DefnSym, OccurrenceAst.Def] = {
```

Comment:
This is legacy style. We should almost never define functions on maps or lists. Just on the individual elements. Then use map/parMap etc.